### ISSUE TITLE
Reproducing Results from Table 3

### ISSUE THREAD

**Issue by tpfeifle** [KaidiXu/GCN_ADV_Train#6]
Hi Kaidi,

Thanks for the amazing work and making the code publicly available!

I am trying to reproduce the results of Table 3 for the natural model (column 1)
Following the readme I trained a natural model and attacked it with varying epsilon.
```python train.py # Train Natural Model```
```python attack.py --model_dir=nat_cora --perturb_ratio=[perturb_ratio]```

The reported test accuracies by the attack.py always lie below the results from Table 3 (first column):

|    | My Experiment Results (Nat-Model) | Table 3 Column 1 (0% defense)
|----|--------|--------|
| **0% attack**  | 0\.820  | 0\.819 |
| **5% attack**  | 0\.684| 0\.721  |
| **10% attack** | 0\.611 | 0\.673 |
| **15% attack** | 0\.540 | 0\.633 |
| **20% attack** | 0\.484 | 0\.598 |

Based on this observation my two questions are:
1) Are the accuracies of Table 3 different from the attack.py test-accuracies? If so, how?
2) Were the default parameters of the code base used in the experiments and might this discrepancy be the reason for the different accuracies?

Tensorflow-Version: 1.13.1

Thanks again for making this code public!

**Comment by KaidiXu** [comment]
Hi Tim,

Thank you for pointing it out. After I checked the code, it's a bug in attack.py, the total edges should be divided by 2 in attack.py.
Our robust training code calls the PGD_attack.py which does not yield this problem.

Now I fixed this bug, and sorry for the confusion, since we only paid more attention to check the correctness of the robust training part and that code also reports the accuracy for convenience.

Best,
Kaidi Xu

**Comment by tpfeifle** [comment]
Hi Kaidi Xu,

Thank you so much for your fast reply! With this fix the natural model performance under attack aligns quite well.

When attacking the robustly trained models I still see a big difference in accuracy compared to the results in Table 3.
I trained the robust models simply with
`python adv_train_pgd.py` 
and only edited the epsilon factor in line 109 to {0.05, 0.1, 0.2} for the different defense models and saved them in separate locations.

The accuracy this outputs after training (for the case "attack eps = defense eps" as ```[adv support] Test Acc``` and as the final ```Test set results accuracy```) and the accuracy from the attack.py attack are now aligned, but both are different from the reported ones in Table 3.

Were different parameters than the ones in the code base used to train the robust models, or am I missing something else? 
In the paper, it is stated e.g. that the robust models were trained with 32 hidden units, while the default parameter in the code is 64.

Below are the test-accuracies for the combinations of perturb_ratio in attack and defense from my experiments:
|    | 0% defense      | 5% defense      | 10% defense      | 20% defense     |
|----|--------|--------|--------|--------|
| **0% attack**  | **0\.82**  | 0\.792 | 0\.784 |0\.781 |
| **5% attack**  | **0\.743** | 0\.735  | 0\.718 |  0\.713 |
| **10% attack** | 0\.684 | **0\.69** | 0\.681 | 0\.0677 |
| **20% attack** | 0\.611 | 0\.625 | **0\.628** | 0\.627 |

Thank you again for your previous reply!

Best,
Tim

**Comment by KaidiXu** [comment]
Hi Tim,

There are some minor differences in attack.py and PGD_attack.py. The most important one is that in attack.py, we randomly sample 20 times in the last step and use the perturbed graph with the best attack performance. However, during training, we only sample once so that may cause higher accuracy (worse attack performance).  Also, we used different splits of the training/val/testing data 5 times and report the mean value.
I saw your results, the tendency looks correct and the differences may be caused by randomness.
You can report the results with either used attack.py or PGD_attack.py as long as consistent in one table.

Best,
Kaidi Xu
________________________________

**Comment by tpfeifle** [comment]
Hi Kaidi Xu,

Thank you for those insights!
I rerun the experiments to make sure the differences are not only caused by randomness, but the robust models still performed worse than the reported results in Table 3, independent of whether attack.py or PGD_attack.py are used.

Except for the 20% attack, the natural model on average even outperforms the robust models (as also seen in the above table of experiment results).

To make sure that this was not caused by a difference in the ```train.py``` and ```adv_train_pgd.py``` parameters I also used ```adv_train_pgd.py``` with 0.0 perturbation_rate to train a natural model, but this still outperformed the 10%-adv-trained model on the 10%-attack, in line with the previously reported results in the above table.

Were any tricks/additional techniques or different code used for the robust training reported in the paper?
As the reported results are very impressive and the projected gradient descent a really elegant approach I would really love to reproduce them.

Best,
Tim

**Comment by KaidiXu** [comment]
Hi Tim,

After I checked the code, I found the reason.
In pure attack methods like CE-PCG and some in existing methods, there is a trick we used and also stated in the paper, "predicted labels (not their ground-truth label) by an independent pre-trained model that can be used during the attack.", although they cannot achieve not 100% accuracy, but it can improve the performance at around 3-6%.
However, it's not naively compatible with conventional robust training, because the number of predicted label (fake test label) is much larger than train label so that may cause the attacking process pay more attention to the test nodes and lead worse attacking results on training data but better attacking results on testing data. That's acceptable in pure attack method, but the training process only feed in the train_label and train_mask so that they cannot learn the information on other node structures.

The solution is simple, just also using the predicted label for robust training.

Sorry for the confusion, the code suffered a huge refactor before release and some flaws may still exist.
I updated the code accordingly and the results should be much better.

Best,
Kaidi Xu
________________________________

**Comment by tpfeifle** [comment]
Hi Kaidi,

Thank you so much for highlighting the idea of using _predicted labels for robust training_! This helped already tremendously.

I have rerun the experiments with the updated code and the performance indeed does increase a lot compared to the model trained only on the training set. It even outperforms the reported numbers of table 3 in the paper.

In the table below are the accuracies for different epsilon in the robust training using `adv_train_pgd.py` and for the natural model using `train.py`.
While the accuracy is nearly constant for the different epsilon we see a clear difference in accuracy between training only on the train data set (train.py) vs using the predicted labels for training (adv_train.pgd.py). Therefore in the current implementation the robustness seems to be a result of the larger set of labels (predicted labels) and quite independent of the epsilon used in the robust training. This is different from table 3 in the paper where the choice of epsilon in the robust training leads to very different results.

Did you also see this pattern in your experiments, or might there be an issue in the current implementation? I am assuming the version before the refactoring is not available anymore?


|   | train.py  | robust eps=0  | robust eps=0.05 | robust eps=0.1 | robust eps=0.2 |
|----|--------|--------|--------|--------|--------|
| **0% attack**  | 0\.81 | 0\.81 | 0\.81 |0\.81 |0\.81 |
| **5% attack**  | 0\.73 | 0\.79  | 0\.79 |  0\.81 |0\.80 |
| **10% attack** | 0\.68 | 0\.77 | 0\.76 | 0\.78 |0\.77 |
| **20% attack** | 0\.61 | 0\.74 | 0\.73 | 0\.73 |0\.70 |

**Comment by tpfeifle** [comment]
Hi Kaidi,

Thank you for your previous insights and adaptations to the code!
Since our last conversation the following three points crystalized as very important:

1) As seen in the table above **the accuracy does not depend on the epsilon used (at least in this implementation)**. This does not align with the results in table 3 of the paper (see screenshot below). 
_I am assuming the version before the refactoring is not available anymore?_

2) The accuracy increases when using the ```adv_train_pgd.py``` instead of the training without robust-samples using ```train.py```. As seen in the table above ```adv_train_pgd.py``` has a higher accuracy even when no robust samples are used(eps=0). This seems to be due to the "self-training" nature of the adv_train_pgd.py code, which uses the predicted labels on all nodes, instead of only the training data set.
**Could you share, whether column 1 in the table 3 of the paper (screenshot below) was produced using the self-training in adv_train_pgd.py or without as in train.py?**
<img src="https://user-images.githubusercontent.com/6023143/108040525-b2cc9500-703d-11eb-93c7-2bbd0a88006f.png " alt="table 3 of PGD paper" width="400px">

3) The attack in the codebase can be made a lot stronger when attacking the ground truth labels instead of the predicted labels ([Current code attacking predicted labels](https://github.com/KaidiXu/GCN_ADV_Train/blob/8ae82cf58745fdd1887ad6aa92b26b050cd06ff8/adv_train_pgd.py#L163)) 


Any insights on the above questions would be wonderful!

Thank you,
Tim
